ssofact.module

```diff
@@ -35,23 +35,216 @@ function ssofact_help($route_name, RouteMatchInterface $route_match) {
+  \Drupal::service('event_dispatcher')->dispatch(SyncUserDataEvent::SYNC, new SyncUserDataEvent($account, $user_info));
+
+  $article_test = \Drupal::service('role_paywall_article_test.manager');
+  $user_data = \Drupal::service('user.data');
+  $uid = $account->id();
+  if (isset($user_info['alfa_purchases']['purchases']) && is_array($user_info['alfa_purchases']['purchases'])) {
+    $purchases = $user_info['alfa_purchases']['purchases'];
+    $today = date('Ymd');
+    $active_purchases = [];
+    $expired_purchases = [];
+    $have_subscriptions = FALSE;
+    $subscription_timestamp = 0;
+    $have_free_trials = FALSE;
+    $free_trial_timestamp = 0;
+    $have_paid_trials = FALSE;
+    $paid_trial_timestamp = 0;
+    $product_labels = [];
+    $customer_status = '';
+    $subscription_product = '';
+    $last_purchase_timestamp = '';
+
+    // Classify purchases.
+    foreach ($purchases as $access) {
+      $purchase = $access['purchase'];
+      if (!is_product($purchase) && !is_free_trial($purchase) && !is_paid_trial($purchase)) {
+        continue;
+      }
+      // Assuming here no subscriptions schedules for the future.
+      if ($purchase['fromDay'] <= $today && $purchase['toDay'] >= $today) {
+        $active_purchases[] = $access['purchase'];
+      }
+      else {
+        $expired_purchases[] = $access['purchase'];
+      }
+    }
+
+    // No active subscriptions but article test active.
+    if (empty($active_purchases) && !$article_test->hasUserAccessToNextTest($account)) {
+      $customer_status = 'trial | article';
+      $last_purchase_timestamp = $article_test->getLastTestTimestamp($account);
+    }
+    // No active subscriptions neither article-test.
+    elseif (empty($active_purchases)) {
+      extract_purchases_info($expired_purchases, $have_free_trials, $have_paid_trials, $have_subscriptions, $product_labels, 'toDay', $free_trial_timestamp, $paid_trial_timestamp, $subscription_timestamp);
+
+      // Had active subscriptions in the past.
+      if ($have_subscriptions) {
+        $subscription_product = implode('+', $product_labels);
+        $customer_status = 'ex | subscriber | ' . $subscription_product;
+        $last_purchase_timestamp = $subscription_timestamp;
+      }
+      // Assuming that free (as is digital) is more relevant for the
+      // customer journey than the paid (print) trials.
+      elseif ($have_free_trials) {
+        $customer_status = 'ex | trial | free';
+        $last_purchase_timestamp = $free_trial_timestamp;
+      }
+      elseif ($have_free_trials) {
+        $customer_status = 'ex | trial | paid';
+        $last_purchase_timestamp = $paid_trial_timestamp;
+      }
+    }
+    // There are active purchases.
+    else {
+      extract_purchases_info($active_purchases, $have_free_trials, $have_paid_trials, $have_subscriptions, $product_labels, 'fromDay', $free_trial_timestamp, $paid_trial_timestamp, $subscription_timestamp);
```

Comment:
1. Function name is not in module namespace.

1. Minor: A long list of parameters passed by reference is a quite weird coding pattern (if it can be called a pattern at all).  Consider to use a single associative array instead.  We can change this later as long as this is internal and no other code calls the function.